### PR TITLE
Introduce plugins and extract resources as a plugin

### PR DESCRIPTION
**Pull request by bobrovnik**
That's just an API preview, so we can discuss it further

### CODE REVIEW

File: src/controllers/loader/index.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+export const createCombinedLoader = (props: Props) => {
+  const { context, resourceData, isStatic } = props;
+  const entryPoint = entryPointsLoader();
+  const resources = resourcesLoader({
+    context,
+    resourceData,
+    isStatic,
+  });
+
+  return combine(entryPoint, resources);
```
Comment (albertogasparin) [inline src/controllers/loader/index.ts:22]: An alternative to this combine API (very redux-like) could be to just pass an array of loaders. I find it easier to digest imo.
It also means that Props are specific to each individual loader, instead of having `resourceData` passed even when entryPoint does not care about it

File: src/controllers/router-store/index.tsx [hunk 2]
```diff
@@ -398,6 +443,33 @@ const actions: AllRouterActions = {
         history[updateType](updatedRelativePath);
       }
     },
+  loadRoute:
```
Comment (albertogasparin) [inline src/controllers/router-store/index.tsx:446]: This action name is a bit confusing as might be interpreted as loading the full route (while instead we just load the loader 😅 . So `loadLoader` is not great either, but maybe `runLoaderLoad` ? 🤔

Comment (MonicaOlejniczak) [reply src/controllers/router-store/index.tsx:446]: I suggested the `loadRoute` name because in the case of EntryPoints (or any custom loader) you load both the component and data. It also does not have to be limited to this, it could load other things within the route (e.g. other components related to the page)

File: src/controllers/loader/index.ts (hunk 1 above)
Comment (MonicaOlejniczak) [reply src/controllers/loader/index.ts:22]: Yeah, that should be fine, as it hides the `combine` as an implementation detail tbh

I'm not sure what you mean about the `resourceData` though, that should be the case with this API anyway.

File: src/common/types.ts [hunk 3]
```diff
@@ -335,3 +335,23 @@ export type FindRouterContextOptions = {
   location: Location;
   basePath?: string;
 };
+
+export type LoaderAPI<T = any> = {
+  onBeforeRouteChange: (params: {
```
[the remark this answers is not in the document]
Comment (MonicaOlejniczak) [reply src/common/types.ts]: Agreed

File: src/common/types.ts [hunk 4]
```diff
@@ -335,3 +335,24 @@ export type FindRouterContextOptions = {
...
+  }) => void;
+  load: (
+    context: RouterContext & {
+      prevLocationContext?: RouterContext;
+    }
+  ) => T;
+  prefetch: (context: RouterContext) => void;
+};
+
+export type Loader<T> = (props: {
```
Comment (MonicaOlejniczak) [inline src/common/types.ts:353]: Name this `ResourceLoader` since these arguments are specific to resources

File: src/controllers/router-store/index.tsx [hunk 5]
```diff
@@ -116,8 +131,13 @@ const actions: AllRouterActions = {
         basePath,
         location: history.location,
         action: history.action,
+        loader,
       });
-      getResourceStore().actions.hydrate({ resourceContext, resourceData });
+
+      if (loader.hydrate) {
```
Comment (MonicaOlejniczak) [inline src/controllers/router-store/index.tsx:137]: If you rebase the latest changes from master, this method should be gone.

File: src/common/types.ts [hunk 6]
```diff
@@ -335,3 +335,24 @@ export type FindRouterContextOptions = {
   location: Location;
   basePath?: string;
 };
+
+export type LoaderAPI<T = Record<string, unknown>> = {
```
Comment (MonicaOlejniczak) [inline src/common/types.ts:339]: I would just call it `Loader`

File: src/controllers/loader/index.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import { resourcesLoader } from '../../resources/loader';
+
+type Props = {
+  context: ResourceStoreContext | undefined;
+  resourceData: any;
+  timeout?: number;
+  isStatic?: boolean;
+};
+
+export const createCombinedLoader = (props: Props) => {
```
Comment (MonicaOlejniczak) [inline src/controllers/loader/index.ts:13]: Was this intended to be part of the public API? Because imo entrypoints shouldn't be in here. I was thinking that the `Router` can take in a `loader` or `loaders` (if you follow Alberto's suggestion) that defaults to the resource loader, making it a non-breaking change. Then consumers that want to combine them just opt into it through the new API. That way you separate the `Props` from `createCombinedLoader` into a `createResourceLoader`

File: src/controllers/router-store/index.tsx [hunk 8]
```diff
@@ -83,8 +91,14 @@ const actions: AllRouterActions = {
         routes,
         location: history.location,
         action: history.action,
+        loader,
       });
-      getResourceStore().actions.hydrate({ resourceContext, resourceData });
+
+      if (loader.hydrate) {
```
Comment (MonicaOlejniczak) [inline src/controllers/router-store/index.tsx:97]: It should still be possible to put this into the resource loader constructor time, and I don't think it's a big deal if you have to construct two separate loaders (one for the server, one for the client)

File: src/__tests__/integration.test.tsx [hunk 9]
```diff
@@ -226,7 +226,7 @@ describe('<Router /> client-side integration tests', () => {
       };
     }
 
-    it('previous data when transitioning to the same route and resource keys', async () => {
+    it.only('previous data when transitioning to the same route and resource keys', async () => {
```
Comment (albertogasparin) [inline src/__tests__/integration.test.tsx:229]: ⚠️

File: src/controllers/router-store/index.tsx [hunk 10]
```diff
@@ -366,6 +345,33 @@ const actions: AllRouterActions = {
...
+        nextContext = findRouterContext(routes, { location, basePath });
+      }
+
+      if (nextContext == null) return;
+      const nextLocationContext = nextContext;
+
+      batch(() => {
+        loader.prefetch(nextLocationContext);
+        // prefetchResources(nextResources, nextLocationContext, {});
+        if (onPrefetch) onPrefetch(nextLocationContext);
```
Comment (albertogasparin) [inline src/controllers/router-store/index.tsx:371]: Now I wonder: should this be moved to a loader too? This could very well be a `webpackLoader` and could handle what `onPrefetch` does in Jira without the need of a specific prop 👀

File: src/common/utils/combine/index.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    beforeLoad: params => {
+      loaders.forEach(loader => {
+        if (loader.beforeLoad !== undefined) {
+          loader.beforeLoad(params);
+        }
+      });
+    },
+
+    load: loadParams => {
+      return loaders.reduce(
```
Comment (albertogasparin) [inline src/common/utils/combine/index.ts:21]: Can you explain to me why this is reduce and the others are foreach?

File: src/controllers/router/index.tsx [hunk 12]
```diff
@@ -59,16 +75,35 @@ Router.requestResources = async ({
...
   bootstrapStore({
     ...bootstrapProps,
     history: history || createMemoryHistory({ initialEntries: [location] }),
+    loader,
   });
 
-  await requestRouteResources({ timeout });
+  // await requestRouteResources({ timeout });
+
+  await loadRoute().resources;
```
Comment (albertogasparin) [inline src/controllers/router/index.tsx:106]: What if I (as a consumer) have another loader that needs to be awaited?
Isn't it better if this is:
```js
await Promise.all(loadRoute())
```
And `loadRoute` is just a map of the calls of all loader `.load()` calls?

File: src/entry-points/use-entry-point/index.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import { useRouter } from '../../controllers/use-router';
+
+export const useEntryPoint = (): any => {
+  const [{ route }] = useRouter();
+
+  if (!route.entryPoint) {
+    return undefined;
+  }
+
+  return { entryPointReference: route?.entryPoint }; // loadEntryPoint(...)
```
Comment (albertogasparin) [inline src/entry-points/use-entry-point/index.ts:10]: `route` is optionally null here but `!route.entryPoint` does not check for that

File: src/common/types.ts [hunk 14]
```diff
@@ -326,3 +326,23 @@ export type FindRouterContextOptions = {
...
+    prevContext: RouterContext;
+    nextContext: RouterContext;
+  }) => void;
+  load: (context: RouterContext, prevContext?: RouterContext) => T;
+  prefetch: (context: RouterContext) => void;
+};
+
+export type CombinedLoader<T = Record<string, any>> = {
+  hydrate: () => void;
+  beforeLoad: (params: {
```
Comment (albertogasparin) [inline src/common/types.ts:342]: Should this follow the same shape as `load` ? 
```
beforeLoad: (prevContext: RouterContext, nextContext: RouterContext ) => void;
```

File: src/common/types.ts [hunk 15]
```diff
@@ -326,3 +326,23 @@ export type FindRouterContextOptions = {
...
+    prevContext: RouterContext;
+    nextContext: RouterContext;
+  }) => void;
+  load: (context: RouterContext, prevContext?: RouterContext) => T;
+  prefetch: (context: RouterContext) => void;
```
Comment (albertogasparin) [inline src/common/types.ts:347]: Is there any value on providing `prevContext` here too? 🤔

File: src/common/types.ts (hunk 14 above)
Comment (MonicaOlejniczak) [reply src/common/types.ts:342]: Yeah, or the other way around is more flexible for future changes.

`beforeLoad({ nextContext, prevContext })`

`Load({ nextContext, prevContext })`

File: src/controllers/router-store/index.tsx (hunk 10 above)
Comment (MonicaOlejniczak) [reply src/controllers/router-store/index.tsx:371]: Yeah, I think it can too. Jira can easily add a custom loader to maintain parity behaviour. It would be a breaking change though but a minor one at that since there wouldn't be too many consumers.

File: src/common/types.ts [hunk 16]
```diff
@@ -326,3 +326,23 @@ export type FindRouterContextOptions = {
...
+  hydrate?: () => void;
+  beforeLoad?: (params: {
+    prevContext: RouterContext;
+    nextContext: RouterContext;
+  }) => void;
+  load: (context: RouterContext, prevContext?: RouterContext) => T;
+  prefetch: (context: RouterContext) => void;
+};
+
+export type CombinedLoader<T = Record<string, any>> = {
```
Comment (MonicaOlejniczak) [inline src/common/types.ts:340]: Should be removed in favour of Loader since they're now the same

File: src/controllers/router-store/test.tsx [hunk 17]
```diff
@@ -142,6 +165,7 @@ describe('RouterStore', () => {
           route: routes[0],
           routes: routes,
           unlisten: expect.any(Function),
+          loader: expect.any(Object),
```
Comment (MonicaOlejniczak) [inline src/controllers/router-store/test.tsx:168]: Should just be `loader,`

File: src/controllers/router/index.tsx [hunk 18]
```diff
@@ -24,6 +26,7 @@ export const Router = ({
   resourceContext,
   resourceData,
   routes,
+  loaders,
```
Comment (MonicaOlejniczak) [inline src/controllers/router/index.tsx:29]: /nit alphabetise props

File: src/controllers/router/index.tsx [hunk 19]
```diff
@@ -33,6 +36,18 @@ export const Router = ({
     };
   }, []);
 
+  const loader = useMemo(() => {
+    if (loaders) return combine(...loaders);
```
Comment (MonicaOlejniczak) [inline src/controllers/router/index.tsx:40]: Shouldn't need to spread `...loaders`

File: src/controllers/router/index.tsx [hunk 20]
```diff
@@ -33,6 +36,18 @@ export const Router = ({
...
+
+    // default 'loaders' fallback for the first relase
+    const resources = createResourcesLoader({
+      context: resourceContext,
+      resourceData,
+    });
+
+    return combine(resources);
```
Comment (MonicaOlejniczak) [inline src/controllers/router/index.tsx:48]: `return resources`

File: .eslintrc.js [hunk 21]
```diff
@@ -47,6 +47,7 @@ module.exports = {
     'react-hooks/exhaustive-deps': 'warn',
     'react-hooks/rules-of-hooks': 'error',
     semi: 'off',
+    'react/jsx-sort-props': 'error',
```
Comment (MonicaOlejniczak) [inline .eslintrc.js:50]: Can you make this a separate pull request?

File: src/common/utils/combine/index.ts [hunk 22]
```diff
@@ -0,0 +1,39 @@
+import type { Plugin } from '../../types';
+
+export const combine = (plugins: Plugin[]): Required<Plugin> => {
+  return {
+    onHydrate: () => {
+      plugins.forEach(plugin => {
+        if (plugin.onHydrate !== undefined) {
```
Comment (albertogasparin) [inline src/common/utils/combine/index.ts:7]: I'd write these as:
```
if (plugin.onHydrate != null) {
```
Just to be more gentle with somebody setting the method to `null`

Comment (albertogasparin) [reply src/common/utils/combine/index.ts:7]: Also another way to specify this "combination" is to have an API like:
```
plugins.run('onHydrate')
```
That would allow us to expand the supported methods without copy pasting these lines for all methods.

**Review by MonicaOlejniczak — CHANGES_REQUESTED** [review]

File: src/controllers/router/index.tsx [hunk 23]
```diff
@@ -52,27 +69,57 @@ export const Router = ({
...
 
   return getResourceStore().actions.getSafeData();
 };
 
+Router.loadRoute = ({ history, plugins, routes }: LoadRouteParams) => {
+  const { bootstrapStore, loadRoute } = getRouterStore().actions;
+
+  bootstrapStore({
+    routes,
+    history: history,
```
Comment (MonicaOlejniczak) [inline src/controllers/router/index.tsx:116]: /nit `history,`

File: src/resources/utils/get-resources-safe-data/index.ts [hunk 24]
```diff
@@ -0,0 +1,5 @@
+import { getResourceStore } from '../../../controllers/resource-store/index';
+
+export const getResourcesSafeData = () => {
```
Comment (MonicaOlejniczak) [inline src/resources/utils/get-resources-safe-data/index.ts:3]: Can we rename this? Maybe `getSerializedResources`?

File: package-lock.json [hunk 25]
```diff
@@ -1,12 +1,12 @@
 {
   "name": "react-resource-router",
-  "version": "0.21.2",
+  "version": "0.21.3-dev-192.0.0.0",
```
Comment (MonicaOlejniczak) [inline package-lock.json:3]: You should revert this

File: src/controllers/router/index.tsx (hunk 12 above)
Comment (MonicaOlejniczak) [reply src/controllers/router/index.tsx:106]: Seems this is handled internally now

**Review by MonicaOlejniczak — APPROVED** [review]

File: examples/hydration-with-plugins/index.tsx [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+    resourceData: data,
+  });
+
+  const App = () => {
+    return (
+      <Router
+        basePath="/hydration-with-plugins"
+        history={myHistory}
+        plugins={[resourcesPlugin]}
+        resourceData={data}
```
Comment (albertogasparin) [inline examples/hydration-with-plugins/index.tsx:53]: This should no longer be here in the plugin example, right?

File: src/common/types.ts [hunk 27]
```diff
@@ -326,3 +326,16 @@ export type FindRouterContextOptions = {
   location: Location;
   basePath?: string;
 };
+
+export type Plugin<T = Record<string, any>> = {
+  onHydrate?: () => void;
```
Comment (albertogasparin) [inline src/common/types.ts:331]: Should we pass the `context` here too? We do have it available...

File: src/resources/plugin/index.ts [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+            route,
+            match,
+            query,
+          },
+          { timeout }
+        ),
+      };
+    },
+    onRoutePrefetch: (context: RouterContext) => {
+      const { route, match, query } = getRouterState();
```
Comment (albertogasparin) [inline src/resources/plugin/index.ts:85]: Isn't this data available on `context: RouterContext` ? `getRouterState` is a sort of private API, so if possible I'd like plugins to not reach into it

File: src/controllers/router-store/index.tsx [hunk 29]
```diff
@@ -102,6 +109,7 @@ const actions: AllRouterActions = {
       const { route, match, query } = getState();
```
Comment (albertogasparin) [inline src/controllers/router-store/index.tsx:109]: What is the future of `requestRouteResources` action? Is this superseded by one of the plugin methods? Can it be deleted?

File: src/controllers/router-store/index.tsx [hunk 30]
```diff
@@ -366,6 +342,32 @@ const actions: AllRouterActions = {
...
+
+      if (nextContext == null) return;
+      const nextLocationContext = nextContext;
+
+      batch(() => {
+        plugins.onRoutePrefetch(nextLocationContext);
+        if (onPrefetch) onPrefetch(nextLocationContext);
+      });
+    },
+  // ...
```
Comment (albertogasparin) [inline src/controllers/router-store/index.tsx:370]: Nit: remove

File: src/controllers/router-store/test.tsx [hunk 31]
```diff
@@ -49,10 +66,18 @@ describe('RouterStore', () => {
...
+        createPlugins({
+          context: props.resourceContext,
+          resourceData: props.resourceData,
+        });
+
       mount(
         <RouterContainer
           history={history}
           isGlobal
+          plugins={plugins}
```
Comment (albertogasparin) [inline src/controllers/router-store/test.tsx:80]: The fact that we have a `plugins` API that is `[plugin, plugin]` and this `plugins` that is the result of `combine([plugin, plugin])` caused me few headaches while reviewing. My brain was expecting `plugins` props to be the same, not change in some places but not in others.

So what about 
- we use `combine` only inside the bootstrap action. So the props/args called `plugins` are always arrays of plugin instances. 
- We can pick a different name for the value in the router store (eg `combinedPlugins`) so it is more obvious that it is no longer the array of plugins.

The result is also that these tests don't need to know about `combine` at all

File: src/controllers/router-store/types.ts [hunk 32]
```diff
@@ -34,6 +35,7 @@ type PrivateStateProperties = {
   onPrefetch?: (routerContext: RouterContext) => void;
   routes: Routes;
   unlisten: UnlistenHistory | null;
+  plugins: Required<Plugin>;
```
Comment (albertogasparin) [inline src/controllers/router-store/types.ts:38]: Let's call this `combinedPlugins` or something similar so it is more clear that is is the "unified" API and not an array.

File: src/controllers/router-store/types.ts [hunk 33]
```diff
@@ -47,6 +49,7 @@ export type ContainerProps = {
   resourceContext?: ResourceStoreContext;
   resourceData?: ResourceStoreData;
   routes: Routes;
+  plugins: Required<Plugin>;
```
Comment (albertogasparin) [inline src/controllers/router-store/types.ts:52]: I'd prefer this to be `Required<Plugin[]>`

**Review by albertogasparin — APPROVED** [review]

**Review by MonicaOlejniczak — DISMISSED** [review]

File: package.json [hunk 34]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-resource-router",
-  "version": "0.21.2",
```
Comment (albertogasparin) [inline package.json:3]: This should be reverted

**Review by albertogasparin — APPROVED** [review]

File: src/controllers/router-store/test.tsx [hunk 35]
```diff
@@ -172,6 +200,32 @@ describe('RouterStore', () => {
...
+        const plugins = [plugin];
+
+        renderRouterContainer({
+          plugins,
+        });
+
+        expect(plugin.hydrate).toBeCalled();
+      });
+
+      it('plugin onRouteLoad is called on initial render', () => {
```
Comment (MonicaOlejniczak) [inline src/controllers/router-store/test.tsx:216]: `routeLoad`

File: src/controllers/router-store/test.tsx [hunk 36]
```diff
@@ -172,6 +200,32 @@ describe('RouterStore', () => {
         });
       });
 
+      it('onHydrate for each plugin is called', () => {
```
Comment (MonicaOlejniczak) [inline src/controllers/router-store/test.tsx:203]: `hydrate`

File: src/common/types.ts [hunk 37]
```diff
@@ -326,3 +326,20 @@ export type FindRouterContextOptions = {
   location: Location;
   basePath?: string;
 };
+
+export type Plugin = {
+  hydrate?: (params: { context: RouterContext }) => void;
```
Comment (MonicaOlejniczak) [inline src/common/types.ts:331]: Do we still need this?

File: src/controllers/router/index.tsx [hunk 38]
```diff
@@ -52,25 +64,31 @@ export const Router = ({
...
  * The entry point for requesting resource data on the server.
  * Pass the result data into the router as a prop in order to hydrate it.
  */
 Router.requestResources = async ({
   location,
   history,
   timeout,
-  ...bootstrapProps
+  routes,
+  resourceContext: context,
```
Comment (MonicaOlejniczak) [inline src/controllers/router/index.tsx:76]: You should add `basePath` and pass it to `invokePluginLoad`

File: src/controllers/router-store/index.tsx [hunk 39]
```diff
@@ -82,9 +81,10 @@ const actions: AllRouterActions = {
         routes,
         location: history.location,
         action: history.action,
+        plugins,
       });
 
-      getResourceStore().actions.hydrate({ resourceContext, resourceData });
+      plugins.forEach(p => p.hydrate?.({ context: routerContext }));
```
Comment (MonicaOlejniczak) [inline src/controllers/router-store/index.tsx:87]: Remove this one

**Review by MonicaOlejniczak — APPROVED** [review]
